### PR TITLE
Update kubeletplugin API for DRA to v1alpha2

### PR DESCRIPTION
**Pull request by klueska**
#### What type of PR is this?
/kind feature
/kind api-change

#### What this PR does / why we need it:
This PR makes the `NodePrepareResources()` and `NodeUnprepareResource()` calls of the kubeletplugin API for DynamicResourceAllocation symmetrical. It wasn't clear how one would use the set of `CDIDevices` passed back in the `NodeUnprepareResource()` of the `v1alpha1` API, and the new API now passes back the full `ResourceHandle` that was originally passed to the `Prepare()` call. Passing the `ResourceHandle` is strictly more informative and a plugin could always (re)derive the set of `CDIDevices` from it.

This is a breaking change, but this release is scheduled to break multiple APIs for DynamicResourceAllocation, so it makes sense to do this now instead of later.

#### Does this PR introduce a user-facing change?

```release-note
ACTION REQUIRED: We have removed support for the v1alpha1 kubeletplugin API of DynamicResourceManagement. All plugins must update to v1alpha2 in order to function properly going forward.
```

### CODE REVIEW

**Comment by klueska** [comment]
/assing @thockin 
Only the last commit is relevant -- the rest are the result of rebasing on not-yet-merged PRs

**Comment by klueska** [comment]
/assign @thockin
Only the last commit is relevant -- the rest are the result of rebasing on not-yet-merged PRs

**Comment by klueska** [comment]
@bart0sh The DRA integration tests are failing with the following. It's not clear what about my changes could cause this though. Maybe you have some insights.:
```
 [FAILED] Timed out after 60.001s.
  prepared claims on host kind-worker
  Expected
      <[]app.ClaimID | len:1, cap:1>: [
          {
              Name: "external-claim",
              UID: "c22dc915-1fde-4107-b378-2f4d1925b348",
          },
      ]
  to be empty
  In [It] at: /home/kklues/kubernetes/test/e2e/dra/dra.go:140 @ 03/13/23 22:56:29.384

  There were additional failures detected.  To view them in detail run ginkgo -vv
------------------------------

Summarizing 1 Failure:
  [FAIL] [sig-node] DRA [Feature:DynamicResourceAllocation] kubelet [It] must unprepare resources for force-deleted pod
```

**Comment by klueska** [comment]
Rebased now that https://github.com/kubernetes/kubernetes/pull/116513 has merged

**Comment by klueska** [comment]
Removing WIP as the e2e tests are now passing

File: pkg/kubelet/cm/dra/manager.go [hunk 1]
```diff
@@ -99,7 +99,7 @@ func (m *ManagerImpl) PrepareResources(pod *v1.Pod) error {
 				// Check if pod is in the ReservedFor for the claim
 				if !resourceclaim.IsReservedForPod(pod, resourceClaim) {
 					return fmt.Errorf("pod %s(%s) is not allowed to use resource claim %s(%s)",
-						pod.Name, pod.UID, podResourceClaim.Name, resourceClaim.UID)
+						pod.Name, pod.UID, claimName, resourceClaim.UID)
```
Comment (pohly) [inline pkg/kubelet/cm/dra/manager.go:102]: This looks like an unrelated fix. It makes sense, though.

File: pkg/kubelet/cm/dra/manager.go [hunk 2]
```diff
@@ -240,19 +240,45 @@ func (m *ManagerImpl) UnprepareResources(pod *v1.Pod) error {
 			continue
 		}
 
+		// Query claim object from the API server
+		resourceClaim, err := m.kubeClient.ResourceV1alpha2().ResourceClaims(pod.Namespace).Get(
```
Comment (pohly) [inline pkg/kubelet/cm/dra/manager.go:244]: Just thinking out loud....

So now the code relies on:
- The ResourceClaim still existing - guaranteed because the claim is still in use.
- Connectivity to the apiserver - will be needed also for updating the pod status, so okay.
- The allocation not being modified - not meant to be done, but not enforced either. Perhaps add that to https://github.com/kubernetes/kubernetes/pull/116332?

File: pkg/kubelet/cm/dra/manager.go [hunk 3]
```diff
@@ -269,7 +295,7 @@ func (m *ManagerImpl) UnprepareResources(pod *v1.Pod) error {
 		m.cache.delete(claimInfo.ClaimName, pod.Namespace)
 
 		// Checkpoint to reduce redundant calls to NodeUnPrepareResource() after a kubelet restart.
-		err := m.cache.syncToCheckpoint()
+		err = m.cache.syncToCheckpoint()
```
Comment (pohly) [inline pkg/kubelet/cm/dra/manager.go:298]: nit:
```
if err := m.cache.syncToCheckpoint(); err != nil {
```

Comment (pohly) [reply pkg/kubelet/cm/dra/manager.go:298]: Also below, although it's perhaps better to not make too many unrelated changes?

File: pkg/kubelet/cm/dra/manager.go (hunk 2 above)
Comment (klueska) [reply pkg/kubelet/cm/dra/manager.go:244]: All of this is true, yes.

File: pkg/kubelet/cm/dra/manager.go (hunk 1 above)
Comment (klueska) [reply pkg/kubelet/cm/dra/manager.go:102]: It’s somewhat unrelated. It does bring this function in line with what is done in the additions to the unprepared call, which is why it felt relevant.

File: pkg/kubelet/cm/dra/manager.go (hunk 3 above)
Comment (klueska) [reply pkg/kubelet/cm/dra/manager.go:298]: Hmm. I usually explicitly tell people to split these in my reviews. With the philosophy that if the function was really long or there were multiple return values we would always split it, and I’d prefer to remain consistent throughout. I know some people argue this is better for variable scoping, but this regiment breaks down as soon as you consider all of the other functions that reuse err as a variable throughout the same function. I prefer to be able to reason about each call site similarly and if you really need a new err variable for your call, you should be using a new name for it and not hiding it in the fact that it’s scoped differently.

File: pkg/kubelet/cm/dra/manager.go (hunk 2 above)
Comment (klueska) [reply pkg/kubelet/cm/dra/manager.go:244]: Validation added to https://github.com/kubernetes/kubernetes/pull/116332 as suggested.

File: pkg/kubelet/cm/dra/manager.go (hunk 3 above)
Comment (klueska) [reply pkg/kubelet/cm/dra/manager.go:298]: I will leave it as is

File: pkg/kubelet/cm/dra/manager.go [hunk 4]
```diff
@@ -130,10 +123,16 @@ func (m *ManagerImpl) PrepareResources(pod *v1.Pod) error {
 
 				// Walk through each resourceHandle
 				for _, resourceHandle := range resourceHandles {
+					// If no KubeletPluginName is provided, we use the DriverName
+					pluginName := resourceHandle.KubeletPluginName
```
Comment (bart0sh) [inline pkg/kubelet/cm/dra/manager.go:127]: Wouldn't empty KubeletPluginName fail a validation in pkg/apis/resource/validation/validation.go ?

**Comment by bart0sh** [comment]
@klueska it's still failing "must unprepare resources for force-deleted pod" test case in my env. I've re-fetched and rebuilt PR branch. Does it fail in your environment as well?

**Comment by klueska** [comment]
> @klueska it's still failing "must unprepare resources for force-deleted pod" test case in my env. I've re-fetched and rebuilt PR branch. Does it fail in your environment as well?

No it does not.

Make sure you redeploy your kubelet (or delete/recrreate the kind cluster that @pohly put together for testing).

**Comment by bart0sh** [comment]
I'm not using kind cluster. I'm running `hacks/local-up-cluster.sh` in one console this way:
```
sudo rm -rf /tmp/kube*
sudo rm -rf /var/run/kubernetes

FEATURE_GATES='DynamicResourceAllocation=true,APIServerTracing=false' RUNTIME_CONFIG=resource.k8s.io/v1alpha2 DNS_ADDON="coredns" CGROUP_DRIVER=systemd ALLOW_PRIVILEGED=1 CONTAINER_RUNTIME_ENDPOINT=unix:///var/run/containerd/containerd.sock LOG_LEVEL=6 API_SECURE_PORT=6444 PATH=$(pwd)/third_party/etcd:$PATH ./hack/local-up-cluster.sh -O

```

and then run tests in another console this way:
```
KUBECONFIG=/var/run/kubernetes/admin.kubeconfig _output/bin/ginkgo -p -v -focus=Feature:DynamicResourceAllocation ./test/e2e
```

**Comment by bart0sh** [comment]
here is how it fails:
```
  STEP: force delete test pod tester-1 @ 03/14/23 19:47:08.12
  STEP: waiting for resources on 127.0.0.1 to be unprepared @ 03/14/23 19:47:08.13
  [FAILED] in [It] - /home/ed/go/src/k8s.io/kubernetes/test/e2e/dra/dra.go:140 @ 03/14/23 19:48:08.132
  Mar 14 19:48:08.132: INFO: Waiting up to 3m0s for all (but 0) nodes to be ready
  STEP: delete pods and claims @ 03/14/23 19:48:08.143
  STEP: deleting *v1alpha2.ResourceClaim dra-9583/external-claim @ 03/14/23 19:48:08.154
  STEP: waiting for resources on 127.0.0.1 to be unprepared @ 03/14/23 19:48:08.161
  [FAILED] in [DeferCleanup (Each)] - /home/ed/go/src/k8s.io/kubernetes/test/e2e/dra/dra.go:805 @ 03/14/23 19:49:08.163
  I0314 19:49:08.163685   12531 controller.go:310] "resource controller: Shutting down" driver="dra-9583.k8s.io"
  E0314 19:49:08.166305   12531 nonblockinggrpcserver.go:101] "kubelet plugin/dra: GRPC server failed" err="listening was stopped" node="127.0.0.1" pod="dra-9583/dra-test-driver-9ldrx"

...

  Mar 14 19:49:08.192: INFO: 
  Logging kubelet events for node 127.0.0.1
  Mar 14 19:49:08.196: INFO: 
  Logging pods the kubelet thinks is on node 127.0.0.1
  Mar 14 19:49:08.200: INFO: Unable to retrieve kubelet pods for node 127.0.0.1: address not allowed
  STEP: Destroying namespace "dra-9583" for this suite. @ 03/14/23 19:49:08.201
  << Timeline

  [FAILED] Timed out after 60.001s.
  prepared claims on host 127.0.0.1
  Expected
      <[]app.ClaimID | len:1, cap:1>: [
          {
              Name: "external-claim",
              UID: "c80066f1-5d72-4f55-9751-3bd70651b0ab",
          },
      ]
  to be empty
  In [It] at: /home/ed/go/src/k8s.io/kubernetes/test/e2e/dra/dra.go:140 @ 03/14/23 19:48:08.132

  There were additional failures detected.  To view them in detail run ginkgo -vv
```

**Comment by klueska** [comment]
I just ran things your way and everything passed. The only difference is that fewer tests seem to run (18 instead of 22). Probably because of different flags when bringing up the cluster via kind, or the fact that the kind cluster is run with multiple worker nodes.

File: pkg/kubelet/cm/dra/manager.go (hunk 4 above)
Comment (klueska) [reply pkg/kubelet/cm/dra/manager.go:127]: no -- it is an optional field by design

**Comment by bart0sh** [comment]
I'm ok with the PR if it works for you. I can try to run the tests again if needed.

File: pkg/kubelet/cm/dra/manager.go (hunk 4 above)
Comment (bart0sh) [reply pkg/kubelet/cm/dra/manager.go:127]: thanks. makes sense to me.

**Comment by thockin** [comment]
Thanks!

/lgtm
/approve
